src/docs.rs

```diff
@@ -153,7 +171,81 @@ pub fn generate_html(
     files
 }
 
-fn function(statement: &TypedStatement) -> Option<Function<'_>> {
+fn build_source_link_generator(
+    project_root: impl AsRef<Path>,
+    project_config: &PackageConfig,
+    module_path: &PathBuf,
+    name: &String,
+) -> Box<dyn SourceLinkGenerator> {
+    match (&project_config.version, &project_config.repository) {
+        (Some(version), Repository::GitHub { url }) => {
+            let relative_path = module_path
+                .strip_prefix(&project_root)
+                .map(|path| path.to_str())
+                .unwrap_or_default()
+                .unwrap_or_default();
+
+            let src = std::fs::read_to_string(&module_path).unwrap_or_default();
+            let codespan_file = codespan_reporting::files::SimpleFile::new(name.clone(), src);
+
+            Box::new(GitHubSourceLinkGenerator {
+                codespan_file,
+                relative_path: relative_path.to_string(),
+                url: url.clone(),
+                version: version.clone(),
+            })
+        }
+        // If we either don't have a version or aren't dealing with a GitHub repository then we
+        // can't generate source code links so we generate empty strings for the urls
+        _ => Box::new(EmptySourceLinkGenerator {}),
+    }
+}
+
+trait SourceLinkGenerator {
```

Comment:
Nitpick: Let's call this `SourceLinks` and move all related code to `src/docs/source_links.rs`

Reply:
Sounds good.